ledger/apply/asset.go

```diff
@@ -208,8 +208,22 @@ func takeOut(balances Balances, addr basics.Address, asset basics.AssetIndex, am
 		return fmt.Errorf("asset %v missing from %v", asset, addr)
 	}
 
-	if sndHolding.Frozen && !bypassFreeze {
-		return fmt.Errorf("asset %v frozen in %v", asset, addr)
+	if !bypassFreeze {
+		if sndHolding.Frozen {
+			return fmt.Errorf("asset %v frozen in %v", asset, addr)
+		}
+
+		assetParams, _, err := getParams(balances, asset)
+		if err != nil {
+			return err
+		}
+		// GlobalFrozen can only be set to true after asset creation, so
+		// it _should_ always have an associated LastAssetFreeze value.
+		// GlobalFrozen is only enforced if the LastAssetFreeze occoured
+		// after an asset holding's LastAccountFreeze.
+		if assetParams.GlobalFrozen && (assetParams.LastAssetFreeze > sndHolding.LastAccountFreeze) {
+			return fmt.Errorf("asset %v globally frozen", asset)
+		}
```

Comment:
It would be nice if we can avoid repeating this logic twice in this file and in the logic package.

Reply:
I've now placed this into its own function `isGlobalFrozen()`, but it's **only** checking the globally frozen side of things, and assumes the local state has already been checked to be false (unfrozen), since that's already checked prior to the consensus upgrade and I didn't want to do a redundant check.